### ISSUE TITLE
VoiceOver occasionally doesn't announce typed characters/numbers in input fields

### ISSUE THREAD

**Issue by mehmetf** [flutter/flutter#71885]
b/174614592

This seems to reproduce with any text field. 

1. Turn on voiceover
2. Type a few characters
3. Delete a few back to back
4. Without waiting for "delete" announcement to finish, type again

Sporadically, you will observe missed announcements for letters. See internal bug for a reproduction of the internal catalog app.

**Comment by chunhtai** [comment]
Flutter currently relying on voice over to pronounce focused soft keyboard key repeat user typing, and that seems to be buggy.

The native ios App relies on some internal mechanism to pronounce the change in textfield. It does not seem to be public documented, I will keep digging.

**Comment by chunhtai** [comment]
I can't really tell how they decide to pronounce the typed character. I have asked a question in stackoverflow https://stackoverflow.com/questions/65225792/voiceover-does-not-repeat-typed-character-on-custom-text-view-with-uitextinput

If we can't figure this out, we may be able to use UIAccessibilityAnnouncementNotification to force voice over to announce the typing.

**Comment by yama3nomori** [comment]
I have the same question.
I tried to add any properties of Textfield() and TextFormField().
But VoiceOver cannot announce characters.

When I tapped a edit box on iOS apps, VoiceOver usually focused the edit box.
But it focus keyboard made by Flutter apps.
I'm thinking at the curse.
But I don't know how to solve it.

**Comment by Hixie** [comment]
Update: We're not sure how to approach this; waiting for feedback from the customer.

**Comment by Hixie** [comment]
@pcsosinski says he will reach out to the customer next week.

**Comment by yasargil** [comment]
We had an accessibility review of our prototype app written with flutter and this bug came up. Except that voiceover never announced a single character typed. For user which rely on that feedback the keyboard is unusable in that state.

**Comment by chunhtai** [comment]
@yasargil thanks that is very helpful, I will look more into it

**Comment by chunhtai** [comment]
yes, I can confirm this is a regression, it works correctly before 1.17.0.

**Comment by chunhtai** [comment]
trace down to engine roll 8cbba0a5c7aaece30af0bbc1406ea2dd5280b222

**Comment by chunhtai** [comment]
cc @xster I can't really tell why the change cause this regression, every thing behind the UItextinput is a black box.

**Comment by chunhtai** [comment]
This is what I have found so far.

In order to send text editing event to the voice over system, it need to be sent through a native UITextInput view. We creates a native UITextInput view in ios embedding to propagate our text editing event, and the flutter text fields uses the same native UITextInput view to sends its text editing events.

for simplicity, lets call native UITextInput view the `NativeField`,  and flutter text field `FlutterField`

The requirement for the `NativeField` to work properly:

1. `NativeField` needs to be attached to the view somewhere.
2. `NativeField` cannot be completely hidden from the accessibility

the (2) is the cause of this bug. Since the previous commit hides the `NativeField` completely, the voiceover does not listen to the its text editing event.

Here are what i have tried so far:

1. We want to hide `NativeField` because both `FlutterField` and `NativeField` will produce accessibility node in voiceover. We want the `FlutterField` to be the only accessibility node in voice over because it contains all the information about rect, label, value.

We cannot completely hide it due to requirement (2), so I have been trying different way to make `NativeField` harder to reach. However, there is always seems to have corners case where the `NativeField` can be reach. That is actually the reason previous commit exist.

2. I tried to always switches the focus back to `FlutterField` as soon as the `NativeField` is focused. This sort of works, but the user will always hears the voiceover pronounces `NativeField` first before the focus switches back to `FlutterField`. it may be confusing to the user.

3. Replaces the `FlutterField` that has focuses with the `NativeField` in the semantics tree. This does not seem possible because `NativeField` is a UIView, and the semantics tree is a Tree of UIAccessibilityElement. The UIView seems to only be able to attach to the tree using [ParentView addSubview:childView]; otherwise, its parent child relationship may be messed up.

It seems like the only way forward is to somehow makes the `FlutterField` implements the UITextInput protocol directly and removes the `NativeField` and hope the voiceover still listens to the `FlutterField`

**Comment by Hixie** [comment]
@chunhtai Is it possible to prototype the solution you describe just enough to tell if it would work?

**Comment by chunhtai** [comment]
I prototyped it, it doesn't work. The soft keyboard only listens to UIView that implements UITextInput. The last suggestion makes UIAccessibilityElement to implement UITextInput, but the soft keyboard does not seem to respond to it.

At this moment, I don't think there is a correct solution.

The best we can have is

(2) always switch the focus back to FlutterField as soon as the NativeField was focused.

and makes NativeField harder to be reached.

**Comment by xster** [comment]
Does moving the native field to the bottom right off-screen not work either?

**Comment by chunhtai** [comment]
no, it does not work, it still receives focus in this scenario

<details>

```dart
import 'package:flutter/material.dart';
import 'package:flutter_driver/flutter_driver.dart';
import 'package:vm_service/vm_service.dart' as vms;

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Counter App',
      home: MyHomePage(title: 'Counter App Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    Navigator.of(context).push(MaterialPageRoute(builder: (_)=>MyHomePage2(title: 'second',) ));
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: TextField(),
      ),
      floatingActionButton: FloatingActionButton(
        // Provide a Key to this button. This allows finding this
        // specific button inside the test suite, and tapping it.
        key: Key('increment'),
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}

class MyHomePage2 extends StatefulWidget {
  MyHomePage2({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePage2State createState() => _MyHomePage2State();
}

class _MyHomePage2State extends State<MyHomePage2> {
  int _counter = 0;

  void _incrementCounter() {
    Navigator.of(context).push(MaterialPageRoute(builder: (_)=>MyHomePage(title: 'second',) ));
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: TextField(autofocus: true,),
      ),
      floatingActionButton: FloatingActionButton(
        // Provide a Key to this button. This allows finding this
        // specific button inside the test suite, and tapping it.
        key: Key('increment'),
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}

class BackButtonListener extends StatefulWidget {
  final Widget child;
  final Future<bool> Function() onBackPressed;

  const BackButtonListener({
    this.child,
    this.onBackPressed,
  });
  @override
  BackButtonListenerState createState()=>BackButtonListenerState();
}

class BackButtonListenerState extends State<BackButtonListener> {

  BackButtonDispatcher dispatcher;
  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    if (dispatcher != null) {
      dispatcher.removeCallback(widget.onBackPressed);
    }
    dispatcher = Router.of(context)
        .backButtonDispatcher
        .createChildBackButtonDispatcher();
    dispatcher.addCallback(widget.onBackPressed);
    dispatcher.takePriority();
  }

  @override
  Widget build(BuildContext context) {
    return widget.child;
  }
}
```

</details>

**Comment by MarcoZehe** [comment]
I am running into this set of problems as well. To me it feels as if the flutter field has the SupportsDirectInteraction trait set. With VoiceOver on, simply touching a text field makes the keyboard come up. Normally, this should only happen with a double tap once the text field has VoiceOver focus (the frame around it). Could that be part of the problem: The Flutter field swallowing gestures it should not?

**Comment by pcsosinski** [comment]
this issue was present in 1.17 and 1.22 correct?

**Comment by chunhtai** [comment]
yes its a regression.

**Comment by pcsosinski** [comment]
we're not planning to pick this into 1.26, the 1.27 beta release will be branching soon.
